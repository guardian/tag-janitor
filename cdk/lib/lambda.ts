import { request } from "https";
import { parse } from "url";
import type { CloudFormationCustomResourceEvent, Context } from "aws-lambda";
import AWS from "aws-sdk";
import type { CustomResourceGetParameterProps } from "./ssmParams";

export function flatten(object: object): Record<string, never> {
  return Object.assign(
    {},
    ...(function _flatten(child: any, path: string[] = []): any {
      return [].concat(
        ...Object.keys(child).map((key) => {
          const childKey = Buffer.isBuffer(child[key]) ? child[key].toString("utf8") : child[key];
          return typeof childKey === "object" && childKey !== null
            ? _flatten(childKey, path.concat([key]))
            : { [path.concat([key]).join(".")]: childKey };
        })
      );
    })(object)
  );
}

export async function handler(event: CloudFormationCustomResourceEvent, context: Context) {
  try {
    console.log(JSON.stringify(event));

    // Default physical resource id
    let physicalResourceId: string;
    switch (event.RequestType) {
      case "Create":
        physicalResourceId = event.LogicalResourceId;
        break;
      case "Update":
        physicalResourceId = event.PhysicalResourceId;
        break;
      case "Delete":
        await respond("SUCCESS", "OK", event.PhysicalResourceId, {});
        return;
    }

    let data: Record<string, string> = {};

    const getParamsProps = event.ResourceProperties.getParamsProps as string | undefined;

    if (getParamsProps) {
      const request = decodeCall(getParamsProps);
      const ssmClient = new AWS.SSM();
      const response = await ssmClient.getParameter(request.apiRequest).promise();
      console.log("Response:", JSON.stringify(response, null, 4));
      data = { ...flatten(response) };
    }

    console.log("data: ", data);
    await respond("SUCCESS", "OK", physicalResourceId, data);
  } catch (e) {
    console.log(e);
    await respond("FAILED", e.message || "Internal Error", context.logStreamName, {});
  }

  function respond(responseStatus: string, reason: string, physicalResourceId: string, data: Record<string, string>) {
    const responseBody = JSON.stringify({
      Status: responseStatus,
      Reason: reason,
      PhysicalResourceId: physicalResourceId,
      StackId: event.StackId,
      RequestId: event.RequestId,
      LogicalResourceId: event.LogicalResourceId,
      NoEcho: false,
      Data: data,
    });

    console.log("Responding", responseBody);
    const parsedUrl = parse(event.ResponseURL);
    const requestOptions = {
      hostname: parsedUrl.hostname,
      path: parsedUrl.path,
      method: "PUT",
      headers: { "content-type": "", "content-length": responseBody.length },
    };

    return new Promise((resolve, reject) => {
      try {
        const r = request(requestOptions, resolve);
        r.on("error", reject);
        r.write(responseBody);
        r.end();
      } catch (e) {
        reject(e);
      }
    });
  }
}

function decodeCall(call: string): CustomResourceGetParameterProps {
  return JSON.parse(call) as CustomResourceGetParameterProps;
}
