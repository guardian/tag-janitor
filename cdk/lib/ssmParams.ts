import { readFileSync } from "fs";
import { join } from "path";
import type { IGrantable, IPrincipal } from "@aws-cdk/aws-iam";
import { Policy, PolicyStatement } from "@aws-cdk/aws-iam";
import { Code, Runtime, SingletonFunction } from "@aws-cdk/aws-lambda";
import type { Reference } from "@aws-cdk/core";
import { Construct, CustomResource, Duration } from "@aws-cdk/core";
import { AwsCustomResourcePolicy } from "@aws-cdk/custom-resources";
import type { GetParameterRequest } from "aws-sdk/clients/ssm";
import type { CdkStack } from "./cdk-stack";

interface GuSSMParameterProps {
  secure?: boolean;
  defaultPath: boolean;
}

export class GuSSMParameter extends Construct implements IGrantable {
  private readonly customResource: CustomResource;
  readonly grantPrincipal: IPrincipal;

  constructor(scope: CdkStack, param: string, props?: GuSSMParameterProps) {
    super(scope, `GuSSMParameter${param}`);
    const filePath = join(__dirname, "lambda.js"); //.replace("/lib/", "/dist/lib/");
    const provider = new SingletonFunction(scope, `${param}Provider`, {
      code: Code.fromInline(readFileSync(filePath).toString()),
      // runtime: new Runtime("nodejs14.x", RuntimeFamily.NODEJS, { supportsInlineCode: true }),
      runtime: Runtime.NODEJS_12_X,
      handler: "index.handler",
      uuid: "eda001a3-b7c8-469d-bc13-787c4e13cfd9",
      lambdaPurpose: "Lambda to fetch SSM parameters",
      timeout: Duration.minutes(2),
    });

    this.grantPrincipal = provider.grantPrincipal;

    const paramArn = `arn:aws:ssm:${scope.region}:${scope.account}:parameter/${param}`;

    const statements: PolicyStatement[] = [
      new PolicyStatement({
        actions: ["ssm:getParameter"],
        resources: AwsCustomResourcePolicy.ANY_RESOURCE, // TODO: Account for stack/stage vs fullPath
      }),
    ];

    const policy = new Policy(scope, `${param}CustomResourcePolicy`, {
      statements: statements,
    });

    if (provider.role !== undefined) {
      policy.attachToRole(provider.role);
    }

    const fullParamPath = `/${scope.stage}/${scope.stack}/${scope.app}/${param}`;

    const getParamsProps: CustomResourceGetParameterProps = {
      apiRequest: {
        Name: props?.defaultPath ? fullParamPath : param,
        WithDecryption: props?.secure,
      },
    };

    this.customResource = new CustomResource(this, `${param}Resource`, {
      resourceType: "Custom::GuGetSSMParameter",
      serviceToken: provider.functionArn,
      pascalCaseProperties: false,
      properties: {
        getParamsProps: JSON.stringify(getParamsProps),
      },
    });

    // If the policy was deleted first, then the function might lose permissions to delete the custom resource
    // This is here so that the policy doesn't get removed before onDelete is called
    this.customResource.node.addDependency(policy);
  }

  public getValueReference(): Reference {
    console.log(this.customResource.toString());
    return this.customResource.getAtt("Parameter.Value");
  }

  public getValue(): string {
    console.log(this.customResource.toString());
    return this.customResource.getAttString("Parameter.Value");
  }
}

export interface CustomResourceGetParameterProps {
  apiRequest: GetParameterRequest;
}

export function GuSSMParameter2(scope: CdkStack, param: string): string {
  return new GuSSMParameter(scope, param, { defaultPath: true }).getValue();
}
