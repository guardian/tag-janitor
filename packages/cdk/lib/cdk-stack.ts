import { Schedule } from '@aws-cdk/aws-events';
import { PolicyStatement } from '@aws-cdk/aws-iam';
import { Runtime, RuntimeFamily } from '@aws-cdk/aws-lambda';
import type { App } from '@aws-cdk/core';
import { Duration } from '@aws-cdk/core';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack, GuStringParameter } from '@guardian/cdk/lib/constructs/core';
import { GuVpc } from '@guardian/cdk/lib/constructs/ec2';
import { GuScheduledLambda } from '@guardian/cdk/lib/patterns/scheduled-lambda';

export class CdkStack extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const app = 'tag-janitor';

		const parameters = {
			topic: new GuStringParameter(this, 'topicArn', {
				description: 'The ARN of the SNS topic to send messages to',
			}),
			accountsAllowList: new GuStringParameter(this, 'accountsAllowList', {
				description: 'A comma separated list of account numbers to include',
			}),
			prismUrl: new GuStringParameter(this, 'prismUrl', {
				description: 'Base URL for Prism',
			}),
		};

		const lambdaFrequency = Duration.days(7);

		// This is copied from the aws-cdk codebase
		// TODO use the enum once updated version of @guardian/cdk
		const node16 = new Runtime('nodejs16.x', RuntimeFamily.NODEJS, {
			supportsInlineCode: true,
		});

		const tagJanitorLambda = new GuScheduledLambda(this, `${app}-lambda`, {
			app: app,
			handler: 'index.handler',
			functionName: `${app}-${this.stage}`,
			runtime: node16,
			fileName: `${app}.zip`,
			environment: {
				STAGE: this.stage,
				TOPIC_ARN: parameters.topic.valueAsString,
				ACCOUNTS_ALLOW_LIST: parameters.accountsAllowList.valueAsString,
				PRISM_URL: parameters.prismUrl.valueAsString,
			},
			description: 'Lambda to notify about instances with missing tags',
			// This lambda needs access to the Deploy Tools VPC so that it can talk to Prism
			vpc: GuVpc.fromIdParameter(this, 'vpc'),
			vpcSubnets: {
				subnets: GuVpc.subnetsFromParameter(this),
			},
			rules: [{ schedule: Schedule.rate(lambdaFrequency) }],
			monitoringConfiguration: {
				alarmName: `High error % from ${app}-${this.stage}`,
				snsTopicName: 'devx-alerts',
				toleratedErrorPercentage: 99,
			},
		});

		tagJanitorLambda.addToRolePolicy(
			new PolicyStatement({
				actions: ['SNS:Publish'],
				resources: [parameters.topic.valueAsString],
			}),
		);
	}
}
