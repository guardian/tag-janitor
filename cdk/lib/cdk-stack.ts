import { Schedule } from "@aws-cdk/aws-events";
import { PolicyStatement } from "@aws-cdk/aws-iam";
import { Runtime } from "@aws-cdk/aws-lambda";
import type { App } from "@aws-cdk/core";
import { Duration } from "@aws-cdk/core";
import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import {
  GuStack,
  GuStringParameter,
  GuSubnetListParameter,
  GuVpcParameter,
} from "@guardian/cdk/lib/constructs/core";
import { GuVpc } from "@guardian/cdk/lib/constructs/ec2";
import { GuLambdaFunction } from "@guardian/cdk/lib/constructs/lambda";

export class CdkStack extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    const parameters = {
      bucketName: new GuStringParameter(this, "BucketName", {
        description: "BucketName",
      }),
      topic: new GuStringParameter(this, "topicArn", {
        description: "The ARN of the SNS topic to send messages to",
      }),
      vpc: new GuVpcParameter(this, "vpcId", {
        description:
          "The VPC ID for the lambda to live in (this allows it to talk to Prism)",
      }), // TODO: Look this up in SSM https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#aws-ssm-parameter-types
      subnets: new GuSubnetListParameter(this, "subnetIds", {
        description:
          "The subnet IDs for the lambda to live in (this allows it to talk to Prism)",
      }), // TODO: Look this up in SSM https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#aws-ssm-parameter-types
      accountsAllowList: new GuStringParameter(this, "accountsAllowList", {
        description: "A comma separated list of account numbers to include",
      }),
      prismUrl: new GuStringParameter(this, "prismUrl", {
        description: "Base URL for Prism",
      }),
    };

    const lambdaFrequency = Duration.days(7);

    const tagJanitorLambda = new GuLambdaFunction(this, `${this.app}-lambda`, {
      handler: "dist/handler.handler",
      functionName: `${this.app}-${this.stage}`,
      runtime: Runtime.NODEJS_12_X,
      code: {
        bucket: parameters.bucketName.valueAsString,
        key: `${this.stack}/${this.stage}/${this.app}/${this.app}.zip`,
      },
      environment: {
        STAGE: this.stage,
        TOPIC_ARN: parameters.topic.valueAsString,
        ACCOUNTS_ALLOW_LIST: parameters.accountsAllowList.valueAsString,
        PRISM_URL: parameters.prismUrl.valueAsString,
      },
      description: "Lambda to notify about instances with missing tags",
      timeout: Duration.seconds(30),
      memorySize: 512,
      vpc: GuVpc.fromId(this, "vpc", parameters.vpc.valueAsString),
      vpcSubnets: {
        subnets: GuVpc.subnets(this, parameters.subnets.valueAsList),
      },
      rules: [
        {
          schedule: Schedule.rate(lambdaFrequency),
          description: `Run tag-janitor every ${lambdaFrequency.toHumanString()}`,
        },
      ],
    });

    tagJanitorLambda.addToRolePolicy(
      new PolicyStatement({
        actions: ["SNS:Publish"],
        resources: [parameters.topic.valueAsString],
      })
    );
  }
}
