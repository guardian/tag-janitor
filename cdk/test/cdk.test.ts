import { SynthUtils } from "@aws-cdk/assert";
import { App } from "@aws-cdk/core";
import { CdkStack } from "../lib/cdk-stack";

describe("The tag-janitor stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const role = new CdkStack(app, "tag-janitor");

    expect(SynthUtils.toCloudFormation(role)).toMatchSnapshot();
  });
});
