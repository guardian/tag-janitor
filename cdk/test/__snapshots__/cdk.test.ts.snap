// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The tag-janitor stack matches the snapshot 1`] = `
Object {
  "Parameters": Object {
    "BucketName": Object {
      "Description": "BucketName",
      "Type": "String",
    },
    "accountsAllowList": Object {
      "Description": "A comma separated list of account numbers to include",
      "Type": "String",
    },
    "prismUrl": Object {
      "Description": "Base URL for Prism",
      "Type": "String",
    },
    "stack": Object {
      "Description": "Stack",
      "Type": "String",
    },
    "stage": Object {
      "Description": "Stage",
      "Type": "String",
    },
    "subnetIds": Object {
      "Description": "The subnet IDs for the lambda to live in (this allows it to talk to Prism)",
      "Type": "List<AWS::EC2::Subnet::Id>",
    },
    "topicArn": Object {
      "Description": "The ARN of the SNS topic to send messages to",
      "Type": "String",
    },
    "vpcId": Object {
      "Description": "The VPC ID for the lambda to live in (this allows it to talk to Prism)",
      "Type": "AWS::EC2::VPC::Id",
    },
  },
  "Resources": Object {
    "ruleF2C1DCDC": Object {
      "Properties": Object {
        "Description": "Run tag-janitor every 7 days",
        "ScheduleExpression": "rate(7 days)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "tagjanitorlambda3E6E11A1",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "tagjanitorlambda3E6E11A1": Object {
      "DependsOn": Array [
        "tagjanitorlambdaServiceRoleDefaultPolicy79766902",
        "tagjanitorlambdaServiceRoleD63E9AA7",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "BucketName",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Ref": "stack",
                },
                "/",
                Object {
                  "Ref": "stage",
                },
                "/tag-janitor/tag-janitor.zip",
              ],
            ],
          },
        },
        "Description": "Lambda to notify about instances with missing tags",
        "Environment": Object {
          "Variables": Object {
            "ACCOUNTS_ALLOW_LIST": Object {
              "Ref": "accountsAllowList",
            },
            "PRISM_URL": Object {
              "Ref": "prismUrl",
            },
            "STAGE": Object {
              "Ref": "stage",
            },
            "TOPIC_ARN": Object {
              "Ref": "topicArn",
            },
          },
        },
        "FunctionName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "tag-janitor-",
              Object {
                "Ref": "stage",
              },
            ],
          ],
        },
        "Handler": "dist/handler.handler",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "tagjanitorlambdaServiceRoleD63E9AA7",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 30,
        "VpcConfig": Object {
          "SecurityGroupIds": Array [
            Object {
              "Fn::GetAtt": Array [
                "tagjanitorlambdaSecurityGroupCDEEB485",
                "GroupId",
              ],
            },
          ],
          "SubnetIds": Object {
            "Ref": "subnetIds",
          },
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "tagjanitorlambdaAllowEventRuletagjanitorruleC323552A3B02531A": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "tagjanitorlambda3E6E11A1",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "ruleF2C1DCDC",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "tagjanitorlambdaSecurityGroupCDEEB485": Object {
      "Properties": Object {
        "GroupDescription": "Automatic security group for Lambda Function tagjanitortagjanitorlambdaBA439B3D",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "tagjanitorlambdaServiceRoleD63E9AA7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "tagjanitorlambdaServiceRoleDefaultPolicy79766902": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "SNS:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "topicArn",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "tagjanitorlambdaServiceRoleDefaultPolicy79766902",
        "Roles": Array [
          Object {
            "Ref": "tagjanitorlambdaServiceRoleD63E9AA7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
